src/dispatch.c

```diff
@@ -92,42 +69,19 @@ SEXP method_(SEXP generic, SEXP signature, SEXP ignore) {
   return m;
 }
 
-SEXP get_class(SEXP object, SEXP envir) {
-    static SEXP fun = NULL;
-    if (fun == NULL) {
-      fun = Rf_findVarInFrame(R_BaseEnv, Rf_install(".class2"));
-    }
-    SEXP call = PROTECT(Rf_lang2(fun, object));
-    SEXP res = Rf_eval(call, envir);
-    UNPROTECT(1);
-    return res;
-}
-
-SEXP object_class_(SEXP object, SEXP envir) {
-  if (Rf_inherits(object, "R7_class")) {
-    return object;
-  }
+SEXP R7_obj_dispatch(SEXP object) {
+  SEXP ns = Rf_findVarInFrame(R_NamespaceRegistry, Rf_install("R7"));
 
-  if (Rf_inherits(object, "R7_object")) {
-    return Rf_getAttrib(object, Rf_install("object_class"));
+  static SEXP obj_dispatch_fun = NULL;
+  if (obj_dispatch_fun == NULL) {
+    obj_dispatch_fun = Rf_findVarInFrame(ns, Rf_install("obj_dispatch"));
   }
 
-  SEXP klass = get_class(object, envir);
+  SEXP methods_extends_call = PROTECT(Rf_lang2(obj_dispatch_fun, object));
+  SEXP res = Rf_eval(methods_extends_call, ns);
+  UNPROTECT(1);
```

Comment:
In the future, you can probably pre-allocate `obj_dispatch_call` at package load time, and then all you'd have to do it set the `object` in the second slot of the call object and `Rf_eval()` it.

You could probably also pre-store the `ns` at package load time as well.

Probably a little premature for all that now though